cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh

```diff
@@ -0,0 +1,488 @@
+                                               pass,
+                                               k);
+  __threadfence();
+
+  if (threadIdx.x == 0) {
+    unsigned int finished = atomicInc(&counter->finished_block_cnt, gridDim.x - 1);
+    isLastBlockDone       = (finished == (gridDim.x - 1));
+  }
+
+  // Synchronize to make sure that each thread reads the correct value of
+  // isLastBlockDone.
+  __syncthreads();
+  if (isLastBlockDone) {
+    if (counter->len == 1 && threadIdx.x == 0) {
+      counter->previous_len = 0;
+      counter->len          = 0;
+    }
+    // init counter, other members of counter is initialized with 0 by
+    // cudaMemset()
+    if (pass == 0 && threadIdx.x == 0) {
+      counter->k   = k;
+      counter->len = len;
+      if (out) { counter->out_back_cnt = 0; }
+    }
+    __syncthreads();
+
+    IdxT ori_k = counter->k;
+
+    if (counter->len > 0) {
+      choose_bucket<T, IdxT, BITS_PER_PASS, NUM_THREAD>(counter, histogram, ori_k);
+    }
+
+    __syncthreads();
+    if (pass == num_passes - 1) {
+      const IdxT previous_len = counter->previous_len;
+      const int want_bucket   = counter->bucket;
+      int start_bit           = calc_start_bit<T, BITS_PER_PASS>(pass);
+      unsigned mask           = calc_mask<T, BITS_PER_PASS>(pass);
+
+      if (!out) {  // radix select
+        for (IdxT i = threadIdx.x; i < previous_len; i += blockDim.x) {
+          const T value = out_buf[i];
+          int bucket    = calc_bucket<T, BITS_PER_PASS>(value, start_bit, mask, greater);
+          if (bucket == want_bucket) {
+            // TODO: UB
+            // could use atomicExch, but it's not defined for T=half
+            counter->kth_value = value;
+            break;
+          }
+        }
+      } else {  // radix topk
+        IdxT& out_cnt = counter->out_cnt;
+        for (IdxT i = threadIdx.x; i < previous_len; i += blockDim.x) {
+          const T value = out_buf[i];
+          int bucket    = calc_bucket<T, BITS_PER_PASS>(value, start_bit, mask, greater);
+          if (bucket < want_bucket) {
+            IdxT pos     = atomicAdd(&out_cnt, IdxT(1));
+            out[pos]     = value;
+            out_idx[pos] = out_idx_buf[i];
+          } else if (bucket == want_bucket) {
+            IdxT needed_num_of_kth = counter->k;
+            IdxT back_pos          = atomicAdd(&(counter->out_back_cnt), IdxT(1));
+            if (back_pos < needed_num_of_kth) {
+              IdxT pos     = k - 1 - back_pos;
+              out[pos]     = value;
+              out_idx[pos] = out_idx_buf[i];
+            }
+          }
+        }
+        __syncthreads();
+      }
+    } else {
+      // reset for next pass
+      for (int i = threadIdx.x; i < num_buckets; i += blockDim.x) {
+        histogram[i] = 0;
+      }
+      if (threadIdx.x == 0) { counter->filter_cnt = 0; }
+    }
+  }
+}
+
+template <typename T, typename IdxT, int BITS_PER_PASS, int NUM_THREAD>
+void radix_topk(const T* in,
```

Comment:
(NB: I removed the other function since the review). The interface now is exactly the same as the public one and the rest implementations in the `detail` namespace. I guess, no need to repeat the same docs multiple times?..